cachi2/interface/cli.py

```diff
@@ -204,11 +195,14 @@ def combine_flags() -> list[str]:
         return flags
 
     parsed_packages = tuple(chain.from_iterable(map(parse_packages, package)))
-    request = Request(
-        source_dir=source,
-        output_dir=output,
-        packages=parsed_packages,
-        flags=combine_flags(),
+    request = parse_user_input(
+        Request.parse_obj,
+        {
+            "source_dir": source,
+            "output_dir": output,
+            "packages": parsed_packages,
+            "flags": combine_flags(),
+        },
```

Comment:
I'm not entirely happy about having to call the `parse_user_input` function to get InvalidInput errors. But it was the best I could come up with, alternatives being:

* make a BaseInput class and override `parse_obj`, `parse_file` and `parse_raw`
  * I didn't tests this one, didn't like it because you could still `__init__` directly and get the raw pydantic error
* ~~make a BaseInput class and override `__init__`~~
  * breaks grouped error reporting (see below)
* override `Request.__init__`
  * works, but why is Request the only input model that gets special treatment?

grouped error reporting:
```
Error: InvalidInput: 2 validation errors for user input
packages -> 0 -> type
  unexpected value; permitted: 'gomod' (given=foo; permitted=('gomod',))
packages -> 1 -> type
  unexpected value; permitted: 'gomod' (given=bar; permitted=('gomod',))
``` 
but if the Package model raises InvalidInput instead of pydantic.ValidationError:
```
Error: InvalidInput: 1 validation error for user input
type
  unexpected value; permitted: 'gomod' (given=foo; permitted=('gomod',))
```

Reply:
Can't we treat this in the CLI `handle_error` decorator level?  You would also catch `pydantic.ValidationError` there, and reraise it as InvalidInput.

Reply:
The thing is `pydantic.ValidationError` is not always invalid input. We also have output models, and invalid output would usually be caused by programmer error. But if we're confident we won't mess it up, maybe that's OK :smile:

Reply:
Not sure if that's much better, but we can do this in the `handle_error` decorator:

```
except pydantic.ValidationError as e:
   if e.model.__module__ == "cachi2.core.models.input":
      msg, details = _present_user_input_error(e)
      raise InvalidInput(f"{msg}\n{details}") from e
   else:
       raise e
```

Reply:
hmm, that feels a little too python-hacky to me

Reply:
I tried that out and eventually reached a solution that felt reasonable.

Now, we override `_BaseInput.__init__` and raise an error that inherits from _both_ InvalidInput and pydantic.ValidationError. That way, pydantic still groups together sub-model validation errors as expected.

Because I still wanted the error to show up as InvalidInput, I came up with the idea of private error classes which report the type name of a non-private parent class.

Let me know if that seems acceptable or too magical :smile:

Reply:
It seems magical to me, but I believe I get it. Will see for how long I'll remember it. :) If the approach in the 6th commit will be the one that is used are you going to place code into the 1st commit and drop it?

Reply:
:+1: yeah putting it in the 1st commit would be better

Reply:
But yeah, if it's too much magic maybe I'll just keep the original approach with having to call a specific function when parsing user input